### PR TITLE
[#IR-308] added submit OTP API and changed generate otp API to now return time to resend in response.

### PR DESCRIPTION
**Pull request by chahat-arora**

### CODE REVIEW

File: apps/user/services/authentication.go [hunk 1]
```diff
@@ -95,52 +95,102 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
 	var otp userModels.OTP
 	err = gormDB.Model(&userModels.OTP{}).Where("user_id = ?", userData.ID).Scan(&otp).Error
 	if err != nil && err != gorm.ErrRecordNotFound {
-		return http.StatusInternalServerError, err
+		return 0, http.StatusInternalServerError, err
+	}
+	// when we don't need mail the OTP.
+	if !identifyData.SendOTP {
+		if err == gorm.ErrRecordNotFound {
+			return 0, http.StatusBadRequest, errors.New("We don't find any generated OTP. Please Generate a new one")
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]:
```suggestion
			return 0, http.StatusBadRequest, errors.New("We didn't find any generated OTP. Please generate a new one")
```

File: apps/user/services/authentication.go [hunk 2]
```diff
@@ -95,52 +95,102 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+		}
+		reSendTime = int((otp.ExpiryAt.Unix() - constants.OTPExpiryTime + constants.OTPReCreationTime) - time.Now().Unix())
+		if reSendTime <= 0 {
+			reSendTime = 0
+		}
+		return reSendTime, http.StatusOK, nil
 	}
+
 	if err != gorm.ErrRecordNotFound {
 		if otp.ExpiryAt.Unix()-constants.OTPExpiryTime+constants.OTPReCreationTime > time.Now().Unix() {
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: `otp.ExpiryAt.Unix() - constants.OTPExpiryTime + constants.OTPReCreationTime` is being used multiple times, please evaluate it once and use it everywhere.

Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: If there is an error, `otp` would be an empty object and this calculation might fail. Please verify and discuss.

File: apps/user/services/authentication.go [hunk 3]
```diff
@@ -95,52 +95,102 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+	var RecoveryData userSerializers.Recover
+	err = c.BindJSON(&RecoveryData)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	gormDB := service.DB
+	var userData userModels.User
+	err = gormDB.Model(&userModels.User{}).Where("email = ?", RecoveryData.Email).Scan(&userData).Error
+	if err != nil {
+		if err == gorm.ErrRecordNotFound {
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Please use gorm's method to check for record not found.

File: apps/user/services/authentication.go [hunk 4]
```diff
@@ -95,52 +95,102 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+	if err != nil && err != gorm.ErrRecordNotFound {
+		return http.StatusInternalServerError, err
+	}
+	if err == gorm.ErrRecordNotFound {
+		return http.StatusBadRequest, errors.New("Didn't find any OTP associated with this email. Please re-generate OTP")
+	}
+	if otp.Code != RecoveryData.OTP {
+		return http.StatusBadRequest, errors.New("Invalid OTP")
+	}
+	if otp.ExpiryAt.Unix() < time.Now().Unix() {
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go:196]: We should have added this in the `OTP` get query only and check for record not found.

File: controllers/user_auth.go [hunk 5]
```diff
@@ -18,6 +18,7 @@ func (ctrl UserAuthController) Routes(r *gin.RouterGroup) {
 	r.GET("/login/", ctrl.Login)
 	r.POST("/login/", ctrl.BasicLogin)
 	r.POST("/identify/", ctrl.Identify)
+	r.POST("/code/", ctrl.Recover)
```
Comment (himanshu-garg-165) [inline controllers/user_auth.go:22]: The endpoint URL should be more intuitive.

File: apps/user/services/authentication.go [hunk 6]
```diff
@@ -95,52 +95,102 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
 
 // Identify ...
 func (service AuthenticationService) Identify(c *gin.Context) (
+	reSendTime int,
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go:100]: Can we use a serializer for `reSendTime`?

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: apps/user/services/authentication.go (hunk 6 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go:100]: we don't need this as there only one field and it's not gonna change.

File: controllers/user_auth.go (hunk 5 above)
Comment (chahat-arora) [reply controllers/user_auth.go:22]: I think it is ok as Facebook also uses this for OTP submission page.

File: apps/user/services/authentication.go (hunk 4 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go:196]: we send different messages for expired OTP and not found OTP.

File: apps/user/services/authentication.go (hunk 3 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

File: apps/user/services/authentication.go (hunk 2 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: There is a check for error in the above line.

Comment (chahat-arora) [reply apps/user/services/authentication.go]: It is used only once.

File: apps/user/services/authentication.go (hunk 1 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

File: apps/user/services/authentication.go [hunk 7]
```diff
@@ -97,53 +97,104 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
 	status int,
 	err error) {
 
 	var identifyData userSerializers.Identify
 	err = c.BindJSON(&identifyData)
 	if err != nil {
-		return http.StatusInternalServerError, err
+		return 0, http.StatusInternalServerError, err
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Return a 400 here.

File: apps/user/services/authentication.go [hunk 8]
```diff
@@ -97,53 +97,104 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
-	// when we don't need OTP to the email.
-	if !identifyData.SendOTP {
-		return http.StatusOK, nil
+		return 0, http.StatusInternalServerError, err
 	}
 	var otp userModels.OTP
 	err = gormDB.Model(&userModels.OTP{}).Where("user_id = ?", userData.ID).Scan(&otp).Error
-	if err != nil && !gorm.IsRecordNotFoundError(err) {
-		return http.StatusInternalServerError, err
+	if err != nil && err != gorm.ErrRecordNotFound {
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Why was it changed from the proper implementation?
```suggestion
	if err != nil && !gorm.IsRecordNotFoundError(err)  {
```

File: apps/user/services/authentication.go [hunk 9]
```diff
@@ -97,53 +97,104 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+		reSendTime = int((otp.ExpiryAt.Unix() - constants.OTPExpiryTime + constants.OTPReCreationTime) - time.Now().Unix())
+		if reSendTime <= 0 {
+			reSendTime = 0
+		}
+		return reSendTime, http.StatusOK, nil
+	}
+
 	// if OTP exists then check its validity.
 	if !gorm.IsRecordNotFoundError(err) {
 		if otp.ExpiryAt.Unix()-constants.OTPExpiryTime+constants.OTPReCreationTime > time.Now().Unix() {
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Please look into it as discussed.

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: apps/user/services/authentication.go (hunk 9 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

File: apps/user/services/authentication.go (hunk 8 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

File: apps/user/services/authentication.go (hunk 7 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

File: apps/user/services/authentication.go [hunk 10]
```diff
@@ -98,58 +97,121 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+	if otp.Code != RecoveryData.OTP {
+		return http.StatusBadRequest, errors.New("Invalid OTP")
+	}
+	if otp.ExpiryAt.Unix() < time.Now().Unix() {
+		return http.StatusBadRequest, errors.New("Your OTP is expired. Please re-generate OTP")
+	}
 	return http.StatusOK, nil
 }
 
+func getReSendTime(otp userModels.OTP) int {
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Make it a model method

File: apps/user/services/authentication.go [hunk 11]
```diff
@@ -98,58 +97,121 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
-		return http.StatusOK, nil
+		return 0, http.StatusInternalServerError, err
 	}
+
 	var otp userModels.OTP
 	err = gormDB.Model(&userModels.OTP{}).Where("user_id = ?", userData.ID).Scan(&otp).Error
-	// when there is some internal DB error.
-	if err != nil && !gorm.IsRecordNotFoundError(err) {
-		return http.StatusInternalServerError, err
+	isOTPExists := !gorm.IsRecordNotFoundError(err)
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Fix this as discussed.

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

**Review by himanshu-garg-165 — APPROVED** [review]

**Pull request merged**